neps/nep-0491.md

```diff
@@ -0,0 +1,359 @@
+account. This balance is only useful for storage staking and otherwise can be
+considered burned.
+
+
+## Motivation
+
+Creating new accounts on chain costs a gas fee and a storage staking fee. The
+more state is added to the account, the higher the storage staking fees. When
+deploying a contract on the account, it can quickly go above 1 NEAR per account.
+
+Some business models are okay with paying that fee for users upfront, just to
+get them onboarded. However, if a business does that today, their users can
+delete their new accounts and spend the tokens intended for storage staking in
+other ways. Since this is free for the user, they are financially incentivized
+to repeat this action for as long as the business has funds left in the faucet.
+
+The protocol should allow to create accounts in a way that is not susceptible to
+such refund abuse. This would at least change the incentives such that creating
+fake users is no longer profitable.
+
+Non-refundable storage staking is a further improvement over
+[NEP-448](https://github.com/near/NEPs/pull/448) (Zero Balance Accounts) which
+addressed the same issue but is limited to 770 bytes per account. By lifting the
+limit, sponsored accounts can be used in combination with smart contracts.
+
+## Specification
+
+Users can opt-in to nonrefundable storage when creating new accounts. For that,
+we extend the `Transfer` action with a new field:
+
+```rust
+struct Transfer {
+    pub deposit: Balance,
+    /// If this flag is set, the balance will be added to the receiver's non-refundable balance.
+    pub nonrefundable: bool,
+}
+```
+
+To create a named account today, the typical pattern is a transaction with
+`CreateAccount`, `Transfer`, and `AddKey`. To make the funds nonrefundable, one
+can use the same pattern with the non-refundable flag set to true:
+
+```json
+"Actions": {
+  "CreateAccount": {},
+  "Transfer": { "deposit": "1000000000000000000000000", "nonrefundable": true },
+  "AddKey": { "public_key": "...", "access_key": "..." }
+}
+```
+
+Adding a second `Transfer` action allows the combination of nonrefundable
+balance and refundable balance. This allows the user to make calls where they
+need to attach balance, for example an FT transfer which requires 1 yocto NEAR.
+
+```json
+"Actions": {
+  "CreateAccount": {},
+  "Transfer": { "deposit": "1000000000000000000000000", "nonrefundable": true },
+  "Transfer": { "deposit": "100", "nonrefundable": false },
+  "AddKey": { "public_key": "...", "access_key": "..." }
+}
+```
+
+To create implicit accounts, the current protocol requires a single `Transfer`
+action without further actions in the same transaction. With this proposal,
+setting it to non-refundable will make the full balance non-refundable. If users
+want to add refundable balance on top, it requires a second transaction.
+
+If a non-refundable transfer arrives at an account that already exists, it will
```

Comment:
Because the non-refundable balance exists only for storage staking and it would be confusing if somebody would top up non-refundable balance and then it all gets burned when accounts gets deleted.

Reply:
See https://github.com/jakmeier/NEPs/blob/non_refundable_storage/neps/nep-0491.md#variation-allow-adding-nonrefundable-balance-to-existing-accounts